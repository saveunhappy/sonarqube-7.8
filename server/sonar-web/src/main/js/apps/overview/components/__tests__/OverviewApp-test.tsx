/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import { OverviewApp } from '../OverviewApp';
import { getMeasuresAndMeta } from '../../../../api/measures';
import { getAllTimeMachineData } from '../../../../api/time-machine';
import {
  mockMainBranch,
  mockComponent,
  mockMetric,
  mockMeasure
} from '../../../../helpers/testMocks';
import { waitAndUpdate } from '../../../../helpers/testUtils';

jest.mock('../../../../api/measures', () => {
  const { mockMeasure, mockMetric } = getMockHelpers();
  return {
    getMeasuresAndMeta: jest.fn().mockResolvedValue({
      component: {
        measures: [mockMeasure({ metric: 'ncloc' }), mockMeasure({ metric: 'coverage' })],
        name: 'foo'
      },
      metrics: [mockMetric({ key: 'ncloc' }), mockMetric()]
    })
  };
});

jest.mock('../../../../helpers/dates', () => ({
  parseDate: jest.fn(date => date)
}));

jest.mock('../../../../api/time-machine', () => ({
  getAllTimeMachineData: jest.fn().mockResolvedValue({
    measures: [
      { metric: 'bugs', history: [{ date: '2019-01-05', value: '2.0' }] },
      { metric: 'vulnerabilities', history: [{ date: '2019-01-05', value: '0' }] },
      { metric: 'sqale_index', history: [{ date: '2019-01-01', value: '1.0' }] },
      { metric: 'duplicated_lines_density', history: [{ date: '2019-01-02', value: '1.0' }] },
      { metric: 'ncloc', history: [{ date: '2019-01-03', value: '10000' }] },
      { metric: 'coverage', history: [{ date: '2019-01-04', value: '95.5' }] }
    ]
  })
}));

it('should render correctly', async () => {
  const wrapper = shallowRender();
  expect(wrapper).toMatchSnapshot();

  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
  expect(getMeasuresAndMeta).toBeCalled();
  expect(getAllTimeMachineData).toBeCalled();
});

it('should render correctly if no measures are found', async () => {
  (getMeasuresAndMeta as jest.Mock).mockResolvedValue({
    component: {
      measures: [mockMeasure({ metric: 'coverage' })],
      name: 'foo'
    },
    metrics: [mockMetric()]
  });
  const wrapper = shallowRender();

  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
});

function getMockHelpers() {
  // We use this little "force-requiring" instead of an import statement in
  // order to prevent a hoisting race condition while mocking. If we want to use
  // a mock helper in a Jest mock, we have to require it like this. Otherwise,
  // we get errors like:
  //     ReferenceError: testMocks_1 is not defined
  return require.requireActual('../../../../helpers/testMocks');
}

function shallowRender(props: Partial<OverviewApp['props']> = {}) {
  return shallow(
    <OverviewApp
      branchLike={mockMainBranch()}
      component={mockComponent({ name: 'foo' })}
      fetchMetrics={jest.fn()}
      metrics={{ coverage: mockMetric() }}
      onComponentChange={jest.fn()}
      {...props}
    />
  );
}
