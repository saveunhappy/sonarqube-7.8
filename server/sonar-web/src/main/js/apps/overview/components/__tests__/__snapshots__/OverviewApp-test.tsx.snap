// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="text-center"
>
  <i
    className="spinner spinner-margin"
  />
</div>
`;

exports[`should render correctly 2`] = `
<div
  className="page page-limited"
>
  <div
    className="overview page-with-sidebar"
  >
    <A11ySkipTarget
      anchor="overview_main"
    />
    <div
      className="overview-main page-main"
    >
      <QualityGate
        branchLike={
          Object {
            "analysisDate": "2018-01-01",
            "isMain": true,
            "name": "master",
          }
        }
        component={
          Object {
            "breadcrumbs": Array [],
            "key": "my-project",
            "name": "foo",
            "organization": "foo",
            "qualifier": "TRK",
            "qualityGate": Object {
              "isDefault": true,
              "key": "30",
              "name": "Sonar way",
            },
            "qualityProfiles": Array [
              Object {
                "deleted": false,
                "key": "my-qp",
                "language": "ts",
                "name": "Sonar way",
              },
            ],
            "tags": Array [],
          }
        }
        measures={
          Array [
            Object {
              "bestValue": true,
              "metric": Object {
                "id": "coverage",
                "key": "ncloc",
                "name": "Coverage",
                "type": "PERCENT",
              },
              "periods": Array [
                Object {
                  "bestValue": true,
                  "index": 1,
                  "value": "1.0",
                },
              ],
              "value": "1.0",
            },
            Object {
              "bestValue": true,
              "metric": Object {
                "id": "coverage",
                "key": "coverage",
                "name": "Coverage",
                "type": "PERCENT",
              },
              "periods": Array [
                Object {
                  "bestValue": true,
                  "index": 1,
                  "value": "1.0",
                },
              ],
              "value": "1.0",
            },
          ]
        }
      />
      <div
        className="overview-domains-list"
      >
        <enhance(Bugs)
          branchLike={
            Object {
              "analysisDate": "2018-01-01",
              "isMain": true,
              "name": "master",
            }
          }
          component={
            Object {
              "breadcrumbs": Array [],
              "key": "my-project",
              "name": "foo",
              "organization": "foo",
              "qualifier": "TRK",
              "qualityGate": Object {
                "isDefault": true,
                "key": "30",
                "name": "Sonar way",
              },
              "qualityProfiles": Array [
                Object {
                  "deleted": false,
                  "key": "my-qp",
                  "language": "ts",
                  "name": "Sonar way",
                },
              ],
              "tags": Array [],
            }
          }
          history={
            Object {
              "bugs": Array [
                Object {
                  "date": "2019-01-05",
                  "value": "2.0",
                },
              ],
              "coverage": Array [
                Object {
                  "date": "2019-01-04",
                  "value": "95.5",
                },
              ],
              "duplicated_lines_density": Array [
                Object {
                  "date": "2019-01-02",
                  "value": "1.0",
                },
              ],
              "ncloc": Array [
                Object {
                  "date": "2019-01-03",
                  "value": "10000",
                },
              ],
              "sqale_index": Array [
                Object {
                  "date": "2019-01-01",
                  "value": "1.0",
                },
              ],
              "vulnerabilities": Array [
                Object {
                  "date": "2019-01-05",
                  "value": "0",
                },
              ],
            }
          }
          historyStartDate="2019-01-05"
          measures={
            Array [
              Object {
                "bestValue": true,
                "metric": Object {
                  "id": "coverage",
                  "key": "ncloc",
                  "name": "Coverage",
                  "type": "PERCENT",
                },
                "periods": Array [
                  Object {
                    "bestValue": true,
                    "index": 1,
                    "value": "1.0",
                  },
                ],
                "value": "1.0",
              },
              Object {
                "bestValue": true,
                "metric": Object {
                  "id": "coverage",
                  "key": "coverage",
                  "name": "Coverage",
                  "type": "PERCENT",
                },
                "periods": Array [
                  Object {
                    "bestValue": true,
                    "index": 1,
                    "value": "1.0",
                  },
                ],
                "value": "1.0",
              },
            ]
          }
        />
        <enhance(VulnerabiltiesAndHotspots)
          branchLike={
            Object {
              "analysisDate": "2018-01-01",
              "isMain": true,
              "name": "master",
            }
          }
          component={
            Object {
              "breadcrumbs": Array [],
              "key": "my-project",
              "name": "foo",
              "organization": "foo",
              "qualifier": "TRK",
              "qualityGate": Object {
                "isDefault": true,
                "key": "30",
                "name": "Sonar way",
              },
              "qualityProfiles": Array [
                Object {
                  "deleted": false,
                  "key": "my-qp",
                  "language": "ts",
                  "name": "Sonar way",
                },
              ],
              "tags": Array [],
            }
          }
          history={
            Object {
              "bugs": Array [
                Object {
                  "date": "2019-01-05",
                  "value": "2.0",
                },
              ],
              "coverage": Array [
                Object {
                  "date": "2019-01-04",
                  "value": "95.5",
                },
              ],
              "duplicated_lines_density": Array [
                Object {
                  "date": "2019-01-02",
                  "value": "1.0",
                },
              ],
              "ncloc": Array [
                Object {
                  "date": "2019-01-03",
                  "value": "10000",
                },
              ],
              "sqale_index": Array [
                Object {
                  "date": "2019-01-01",
                  "value": "1.0",
                },
              ],
              "vulnerabilities": Array [
                Object {
                  "date": "2019-01-05",
                  "value": "0",
                },
              ],
            }
          }
          historyStartDate="2019-01-05"
          measures={
            Array [
              Object {
                "bestValue": true,
                "metric": Object {
                  "id": "coverage",
                  "key": "ncloc",
                  "name": "Coverage",
                  "type": "PERCENT",
                },
                "periods": Array [
                  Object {
                    "bestValue": true,
                    "index": 1,
                    "value": "1.0",
                  },
                ],
                "value": "1.0",
              },
              Object {
                "bestValue": true,
                "metric": Object {
                  "id": "coverage",
                  "key": "coverage",
                  "name": "Coverage",
                  "type": "PERCENT",
                },
                "periods": Array [
                  Object {
                    "bestValue": true,
                    "index": 1,
                    "value": "1.0",
                  },
                ],
                "value": "1.0",
              },
            ]
          }
        />
        <enhance(CodeSmells)
          branchLike={
            Object {
              "analysisDate": "2018-01-01",
              "isMain": true,
              "name": "master",
            }
          }
          component={
            Object {
              "breadcrumbs": Array [],
              "key": "my-project",
              "name": "foo",
              "organization": "foo",
              "qualifier": "TRK",
              "qualityGate": Object {
                "isDefault": true,
                "key": "30",
                "name": "Sonar way",
              },
              "qualityProfiles": Array [
                Object {
                  "deleted": false,
                  "key": "my-qp",
                  "language": "ts",
                  "name": "Sonar way",
                },
              ],
              "tags": Array [],
            }
          }
          history={
            Object {
              "bugs": Array [
                Object {
                  "date": "2019-01-05",
                  "value": "2.0",
                },
              ],
              "coverage": Array [
                Object {
                  "date": "2019-01-04",
                  "value": "95.5",
                },
              ],
              "duplicated_lines_density": Array [
                Object {
                  "date": "2019-01-02",
                  "value": "1.0",
                },
              ],
              "ncloc": Array [
                Object {
                  "date": "2019-01-03",
                  "value": "10000",
                },
              ],
              "sqale_index": Array [
                Object {
                  "date": "2019-01-01",
                  "value": "1.0",
                },
              ],
              "vulnerabilities": Array [
                Object {
                  "date": "2019-01-05",
                  "value": "0",
                },
              ],
            }
          }
          historyStartDate="2019-01-05"
          measures={
            Array [
              Object {
                "bestValue": true,
                "metric": Object {
                  "id": "coverage",
                  "key": "ncloc",
                  "name": "Coverage",
                  "type": "PERCENT",
                },
                "periods": Array [
                  Object {
                    "bestValue": true,
                    "index": 1,
                    "value": "1.0",
                  },
                ],
                "value": "1.0",
              },
              Object {
                "bestValue": true,
                "metric": Object {
                  "id": "coverage",
                  "key": "coverage",
                  "name": "Coverage",
                  "type": "PERCENT",
                },
                "periods": Array [
                  Object {
                    "bestValue": true,
                    "index": 1,
                    "value": "1.0",
                  },
                ],
                "value": "1.0",
              },
            ]
          }
        />
        <enhance(Coverage)
          branchLike={
            Object {
              "analysisDate": "2018-01-01",
              "isMain": true,
              "name": "master",
            }
          }
          component={
            Object {
              "breadcrumbs": Array [],
              "key": "my-project",
              "name": "foo",
              "organization": "foo",
              "qualifier": "TRK",
              "qualityGate": Object {
                "isDefault": true,
                "key": "30",
                "name": "Sonar way",
              },
              "qualityProfiles": Array [
                Object {
                  "deleted": false,
                  "key": "my-qp",
                  "language": "ts",
                  "name": "Sonar way",
                },
              ],
              "tags": Array [],
            }
          }
          history={
            Object {
              "bugs": Array [
                Object {
                  "date": "2019-01-05",
                  "value": "2.0",
                },
              ],
              "coverage": Array [
                Object {
                  "date": "2019-01-04",
                  "value": "95.5",
                },
              ],
              "duplicated_lines_density": Array [
                Object {
                  "date": "2019-01-02",
                  "value": "1.0",
                },
              ],
              "ncloc": Array [
                Object {
                  "date": "2019-01-03",
                  "value": "10000",
                },
              ],
              "sqale_index": Array [
                Object {
                  "date": "2019-01-01",
                  "value": "1.0",
                },
              ],
              "vulnerabilities": Array [
                Object {
                  "date": "2019-01-05",
                  "value": "0",
                },
              ],
            }
          }
          historyStartDate="2019-01-05"
          measures={
            Array [
              Object {
                "bestValue": true,
                "metric": Object {
                  "id": "coverage",
                  "key": "ncloc",
                  "name": "Coverage",
                  "type": "PERCENT",
                },
                "periods": Array [
                  Object {
                    "bestValue": true,
                    "index": 1,
                    "value": "1.0",
                  },
                ],
                "value": "1.0",
              },
              Object {
                "bestValue": true,
                "metric": Object {
                  "id": "coverage",
                  "key": "coverage",
                  "name": "Coverage",
                  "type": "PERCENT",
                },
                "periods": Array [
                  Object {
                    "bestValue": true,
                    "index": 1,
                    "value": "1.0",
                  },
                ],
                "value": "1.0",
              },
            ]
          }
        />
        <enhance(Duplications)
          branchLike={
            Object {
              "analysisDate": "2018-01-01",
              "isMain": true,
              "name": "master",
            }
          }
          component={
            Object {
              "breadcrumbs": Array [],
              "key": "my-project",
              "name": "foo",
              "organization": "foo",
              "qualifier": "TRK",
              "qualityGate": Object {
                "isDefault": true,
                "key": "30",
                "name": "Sonar way",
              },
              "qualityProfiles": Array [
                Object {
                  "deleted": false,
                  "key": "my-qp",
                  "language": "ts",
                  "name": "Sonar way",
                },
              ],
              "tags": Array [],
            }
          }
          history={
            Object {
              "bugs": Array [
                Object {
                  "date": "2019-01-05",
                  "value": "2.0",
                },
              ],
              "coverage": Array [
                Object {
                  "date": "2019-01-04",
                  "value": "95.5",
                },
              ],
              "duplicated_lines_density": Array [
                Object {
                  "date": "2019-01-02",
                  "value": "1.0",
                },
              ],
              "ncloc": Array [
                Object {
                  "date": "2019-01-03",
                  "value": "10000",
                },
              ],
              "sqale_index": Array [
                Object {
                  "date": "2019-01-01",
                  "value": "1.0",
                },
              ],
              "vulnerabilities": Array [
                Object {
                  "date": "2019-01-05",
                  "value": "0",
                },
              ],
            }
          }
          historyStartDate="2019-01-05"
          measures={
            Array [
              Object {
                "bestValue": true,
                "metric": Object {
                  "id": "coverage",
                  "key": "ncloc",
                  "name": "Coverage",
                  "type": "PERCENT",
                },
                "periods": Array [
                  Object {
                    "bestValue": true,
                    "index": 1,
                    "value": "1.0",
                  },
                ],
                "value": "1.0",
              },
              Object {
                "bestValue": true,
                "metric": Object {
                  "id": "coverage",
                  "key": "coverage",
                  "name": "Coverage",
                  "type": "PERCENT",
                },
                "periods": Array [
                  Object {
                    "bestValue": true,
                    "index": 1,
                    "value": "1.0",
                  },
                ],
                "value": "1.0",
              },
            ]
          }
        />
      </div>
    </div>
    <div
      className="overview-sidebar page-sidebar-fixed"
    >
      <Connect(Meta)
        branchLike={
          Object {
            "analysisDate": "2018-01-01",
            "isMain": true,
            "name": "master",
          }
        }
        component={
          Object {
            "breadcrumbs": Array [],
            "key": "my-project",
            "name": "foo",
            "organization": "foo",
            "qualifier": "TRK",
            "qualityGate": Object {
              "isDefault": true,
              "key": "30",
              "name": "Sonar way",
            },
            "qualityProfiles": Array [
              Object {
                "deleted": false,
                "key": "my-qp",
                "language": "ts",
                "name": "Sonar way",
              },
            ],
            "tags": Array [],
          }
        }
        history={
          Object {
            "bugs": Array [
              Object {
                "date": "2019-01-05",
                "value": "2.0",
              },
            ],
            "coverage": Array [
              Object {
                "date": "2019-01-04",
                "value": "95.5",
              },
            ],
            "duplicated_lines_density": Array [
              Object {
                "date": "2019-01-02",
                "value": "1.0",
              },
            ],
            "ncloc": Array [
              Object {
                "date": "2019-01-03",
                "value": "10000",
              },
            ],
            "sqale_index": Array [
              Object {
                "date": "2019-01-01",
                "value": "1.0",
              },
            ],
            "vulnerabilities": Array [
              Object {
                "date": "2019-01-05",
                "value": "0",
              },
            ],
          }
        }
        measures={
          Array [
            Object {
              "bestValue": true,
              "metric": Object {
                "id": "coverage",
                "key": "ncloc",
                "name": "Coverage",
                "type": "PERCENT",
              },
              "periods": Array [
                Object {
                  "bestValue": true,
                  "index": 1,
                  "value": "1.0",
                },
              ],
              "value": "1.0",
            },
            Object {
              "bestValue": true,
              "metric": Object {
                "id": "coverage",
                "key": "coverage",
                "name": "Coverage",
                "type": "PERCENT",
              },
              "periods": Array [
                Object {
                  "bestValue": true,
                  "index": 1,
                  "value": "1.0",
                },
              ],
              "value": "1.0",
            },
          ]
        }
        metrics={
          Object {
            "coverage": Object {
              "id": "coverage",
              "key": "coverage",
              "name": "Coverage",
              "type": "PERCENT",
            },
          }
        }
        onComponentChange={[MockFunction]}
      />
    </div>
  </div>
</div>
`;

exports[`should render correctly if no measures are found 1`] = `
<div
  className="page page-limited"
>
  <div
    className="overview page-with-sidebar"
  >
    <A11ySkipTarget
      anchor="overview_main"
    />
    <div
      className="overview-main page-main"
    >
      <h3>
        overview.project.empty
      </h3>
    </div>
    <div
      className="overview-sidebar page-sidebar-fixed"
    >
      <Connect(Meta)
        branchLike={
          Object {
            "analysisDate": "2018-01-01",
            "isMain": true,
            "name": "master",
          }
        }
        component={
          Object {
            "breadcrumbs": Array [],
            "key": "my-project",
            "name": "foo",
            "organization": "foo",
            "qualifier": "TRK",
            "qualityGate": Object {
              "isDefault": true,
              "key": "30",
              "name": "Sonar way",
            },
            "qualityProfiles": Array [
              Object {
                "deleted": false,
                "key": "my-qp",
                "language": "ts",
                "name": "Sonar way",
              },
            ],
            "tags": Array [],
          }
        }
        history={
          Object {
            "bugs": Array [
              Object {
                "date": "2019-01-05",
                "value": "2.0",
              },
            ],
            "coverage": Array [
              Object {
                "date": "2019-01-04",
                "value": "95.5",
              },
            ],
            "duplicated_lines_density": Array [
              Object {
                "date": "2019-01-02",
                "value": "1.0",
              },
            ],
            "ncloc": Array [
              Object {
                "date": "2019-01-03",
                "value": "10000",
              },
            ],
            "sqale_index": Array [
              Object {
                "date": "2019-01-01",
                "value": "1.0",
              },
            ],
            "vulnerabilities": Array [
              Object {
                "date": "2019-01-05",
                "value": "0",
              },
            ],
          }
        }
        measures={
          Array [
            Object {
              "bestValue": true,
              "metric": Object {
                "id": "coverage",
                "key": "coverage",
                "name": "Coverage",
                "type": "PERCENT",
              },
              "periods": Array [
                Object {
                  "bestValue": true,
                  "index": 1,
                  "value": "1.0",
                },
              ],
              "value": "1.0",
            },
          ]
        }
        metrics={
          Object {
            "coverage": Object {
              "id": "coverage",
              "key": "coverage",
              "name": "Coverage",
              "type": "PERCENT",
            },
          }
        }
        onComponentChange={[MockFunction]}
      />
    </div>
  </div>
</div>
`;
